Don't cache and don't forward hop-by-hop headers

ikoveshnikov opened the pull request:
Fix issue #409 

Main changes:
- Update http parser to save _Connection_ header string as list of header names tagged with flag `TFW_STR_VALUE`. Most common tokens `keep-alive` and `close` and also delimiters are not marked with this flag: first processed as special header as most common one, others don't point to any headers.  `TFW_HTTP_CONN_EXTRA` flag for whole message is set if _Connection_ contains more than just `keep-alive` or `close`
- `Keep-alive` header is set as special. That helps not to slow down performance by multiple string comparisons if `Keep-alive` is the only header set by _Connection_. Corresponding unit test was updated.
- Remove hop-by-hop headers when forwarding messages from client to server and vice-versa
- Don't cache hop-by-hop headers for server responses.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -811,3 +815,105 @@ tfw_http_msg_alloc(int type)
+				idxs[id - TFW_HTTP_HDR_RAW] = 1;
+			}
+		}
+	}
```
This whole code block looks every complicated and has code duplication. I believe it should be structured in a different way. An outer loop is needed here. Inside that loop any unneeded chunks are skipped. When we step on a VALUE chunk then start collecting a token in a inner loop. After the inner loop has finished, then the colon is appended, the lookup is done, and the header is marked if found.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -811,3 +815,105 @@ tfw_http_msg_alloc(int type)
+/**
+ * Remove Hop-by-Hop headers in terms of RFC 7230 6.1
+ */
+int tfw_http_rm_hbh_hdrs(TfwHttpMsg *hm)
```
As noted by @ikoveshnikov, these functions should be named `tfw_http_msg_*_hbh_hdrs()` as they reside in `http_msg.c` module.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -811,3 +815,105 @@ tfw_http_msg_alloc(int type)
+		if (hbh_heades[i]) {
+			r = __hdr_del(hm, i + TFW_HTTP_HDR_RAW);
+			if (unlikely(r))
+				goto out;
```
Why not `break` from the loop?

keshonok commented on tempesta_fw/http_parser.c:
```diff
@@ -877,6 +895,12 @@ __FSM_STATE(RGen_OWS) {							\
+ * tokens. If Connection header contains other tokens, message flag
+ * TFW_HTTP_CONN_EXTRA is set to indicate that hop-by-hop headers are set
+ * and must be processed later. Each value will be placed to its chunk marked
+ * with TFW_STR_VALUE flag.
```
Each chunk of a token other than those mentioned above is marked with `TFW_STR_VALUE` flag.

keshonok commented on tempesta_fw/http_parser.c:
```diff
@@ -907,19 +931,25 @@ __parse_connection(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	 */
 	__FSM_STATE(I_ConnOther) {
-		__FSM_I_MATCH_MOVE(token, I_ConnOther);
+		msg->flags |= TFW_HTTP_CONN_EXTRA;
```
This statement will be executed each time another chunk is seen in the input, even it it's just a single byte. Perhaps, an extra state is justified here that designates the start of a token.

ikoveshnikov commented on tempesta_fw/cache.c:
```diff
@@ -846,18 +858,29 @@ tfw_cache_add(TfwHttpResp *resp, TfwHttpReq *req, tfw_http_cache_cb_t action)
+	if (resp->flags & TFW_HTTP_CONN_EXTRA) {
+		/* Fill hop-by-hop identifiers for RAW headers */
+		hbh_hds_raw = kcalloc(resp->h_tbl->off - TFW_HTTP_HDR_RAW,
+				      sizeof(unsigned int), GFP_ATOMIC);
```
Using `TfwStr{}->flags` to indicate hbh header sounds like a great idea.

ikoveshnikov commented:
Changes: 
- Rebased to current master
- Hop-by-hop headers refactored: get rid of memory allocations and VLAs, simplify code, don't try to find hop-by-hop headers twice for same response (during caching and during adjusting)
- removed extra empty lines
- update function documentation

TODO:
- check unit tests (fail on master branch, not caused by the PR)

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -819,3 +848,89 @@ tfw_http_msg_alloc(int type)
+	 * message. Since marks stay in http message until it would be
+	 * destroyed, we can reuse marks later without heavy processing.
+	 */
+	hm->flags &= ~((unsigned int)TFW_HTTP_CONN_EXTRA);
```
This serves as an indicator that allows to NOT perform heavy marking again when it had been done already. However, why not avoid repeated calls to mark the headers altogether? That can be done by setting a flag once the marking is done.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -819,3 +848,89 @@ tfw_http_msg_alloc(int type)
+	__mark_hbh_header(&hm->h_tbl->tbl[TFW_HTTP_HDR_CONNECTION]);
+	__mark_hbh_header(&hm->h_tbl->tbl[TFW_HTTP_HDR_KEEP_ALIVE]);
+
+	if (!(hm->flags & TFW_HTTP_CONN_EXTRA))
```
Please see the comment below regarding repeated calls to this function.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -819,3 +848,89 @@ tfw_http_msg_alloc(int type)
+{
+	int r = 0, i;
+
+	tfw_http_msg_mark_hbh_hdrs(hm);
```
As per above comment, this should be called only when the headers have not been marked yet.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -155,22 +156,50 @@ __hdr_is_singular(const TfwStr *hdr)
 			h = TFW_STR_CHUNK(h, 0);
-		if (tfw_stricmpspn(hdr, h, ':'))
-			continue;
+		if (!tfw_stricmpspn(hdr, h, ':'))
```
Frankly, this doesn't match the strings correctly. It may successfully match strings like `Keep-Alive:` and `Keep-Alive-Forever:`. Why not fix it and the enforce the correct matching?

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -155,22 +156,50 @@ __hdr_is_singular(const TfwStr *hdr)
+			h = TFW_STR_CHUNK(h, 0);
+		if (tfw_str_eq(name, h, name->len)
+		    && (':' == tfw_str_at_index(h, name->len)))
+			break;
```
Please see the comment above. It looks to me that there's no need to introduce two new TwfStr{} strings function `tfw_str_eq()` and `tfw_str_at_index()`. At least not until that functionality is really needed.

The core of this function should be made a separate function that can be used by `__hdr_lookup()`. That separate function would return the `id` if it's found, and it can be used directly in place of this `__hdr_lookup_hdr_name()`.

keshonok commented on tempesta_fw/http_msg.c:
```diff
@@ -819,3 +848,89 @@ tfw_http_msg_alloc(int type)
+				TFW_HTTP_HDR_CONNECTION, &conn_h);
+
+	TFW_STR_FOR_EACH_CHUNK(c, &conn_h, end) {
+		if (c->flags & TFW_STR_VALUE) {
```
Deep nesting, more difficult to read and comprehend the logic.
```C
if (!(c->flags & TFW_STR_VALUE))
        continue;
```

krizhanovsky commented on tempesta_fw/str.c:
```diff
@@ -318,6 +318,67 @@ tfw_stricmpspn(const TfwStr *s1, const TfwStr *s2, int stop)
+ * substring separately.
+ *
+ * @return true if the strings are equal and false otherwise.
+ */
```
It seems both the string functions are unnecessary, see comments above: you can just create temporary compound TfwStr with ":" at the end chunk and use `tfw_stricmpspn()`.

ikoveshnikov commented:
redisigned in #650
